drivers/usb/udc/udc_smartbond.c

```diff
@@ -0,0 +1,1749 @@
+							 uint8_t ep_idx)
+{
+	const struct udc_smartbond_config *config = data->dev->config;
+
+	return ep_idx < config->num_of_eps ? &data->ep_state[1][ep_idx] : NULL;
+}
+
+static void set_nfsr(struct usb_smartbond_data *data, uint8_t val)
+{
+	data->nfsr = val;
+	/*
+	 * Write only lower 2 bits to register, higher bits are used
+	 * to count down till OPERATIONAL state can be entered when
+	 * remote wakeup activated.
+	 */
+	USB->USB_NFSR_REG = val & 3;
+}
+
+static void fill_tx_fifo(struct smartbond_ep_state *ep_state)
+{
+	int remaining;
+	const uint8_t *src;
+	struct smartbond_ep_reg_set *regs = ep_state->regs;
+	struct net_buf *buf = ep_state->buf;
+	const struct udc_ep_config *const ep_cfg = &ep_state->config;
+	const uint16_t mps = udc_mps_ep_size(ep_cfg);
+	const uint8_t ep_idx = USB_EP_GET_IDX(ep_cfg->addr);
+
+	src = buf->data;
+	remaining = buf->len;
+	if (remaining > mps - ep_state->last_packet_size) {
+		remaining = mps - ep_state->last_packet_size;
+	}
+
+	/*
+	 * Loop checks TCOUNT all the time since this value is saturated to 31
+	 * and can't be read just once before.
+	 */
+	while ((regs->txs & USB_USB_TXS1_REG_USB_TCOUNT_Msk) > 0 && remaining > 0) {
+		regs->txd = *src++;
+		ep_state->last_packet_size++;
+		remaining--;
+	}
+
+	if (ep_idx != 0) {
+		if (remaining > 0) {
+			/*
+			 * Max packet size is set to value greater then FIFO.
+			 * Enable fifo level warning to handle larger packets.
+			 */
+			regs->txc |= (3 << USB_USB_TXC1_REG_USB_TFWL_Pos);
+			USB->USB_FWMSK_REG |=
+				BIT(ep_idx - 1 + USB_USB_FWMSK_REG_USB_M_TXWARN31_Pos);
+		} else {
+			regs->txc &= ~USB_USB_TXC1_REG_USB_TFWL_Msk;
+			USB->USB_FWMSK_REG &=
+				~(BIT(ep_idx - 1 + USB_USB_FWMSK_REG_USB_M_TXWARN31_Pos));
+			/* Whole packet already in fifo, no need to
+			 * refill it later.  Mark last.
+			 */
+			regs->txc |= USB_USB_TXC1_REG_USB_LAST_Msk;
+		}
+	}
+}
+
+static bool try_allocate_dma(struct usb_smartbond_data *data, struct smartbond_ep_state *ep_state)
+{
+	struct udc_ep_config *const ep_cfg = &ep_state->config;
+	const uint8_t ep = ep_cfg->addr;
+	uint8_t ep_idx = USB_EP_GET_IDX(ep);
+	uint8_t dir_ix = USB_EP_DIR_IS_OUT(ep) ? 0 : 1;
+
+	if (atomic_ptr_cas(&data->dma_ep[dir_ix], NULL, ep_state)) {
+		if (dir_ix == 0) {
+			USB->USB_DMA_CTRL_REG =
+				(USB->USB_DMA_CTRL_REG & ~USB_USB_DMA_CTRL_REG_USB_DMA_RX_Msk) |
+				((ep_idx - 1) << USB_USB_DMA_CTRL_REG_USB_DMA_RX_Pos);
+		} else {
+			USB->USB_DMA_CTRL_REG =
+				(USB->USB_DMA_CTRL_REG & ~USB_USB_DMA_CTRL_REG_USB_DMA_TX_Msk) |
+				((ep_idx - 1) << USB_USB_DMA_CTRL_REG_USB_DMA_TX_Pos);
+		}
+		USB->USB_DMA_CTRL_REG |= USB_USB_DMA_CTRL_REG_USB_DMA_EN_Msk;
+		return true;
+	} else {
+		return false;
+	}
+}
+
+static void start_rx_dma(const struct usb_smartbond_dma_config *dma_cfg, volatile void *src,
+			 void *dst, uint16_t size)
+{
+	if (dma_reload(dma_cfg->rx_dev, dma_cfg->rx_chan, (uint32_t)src, (uint32_t)dst, size) < 0) {
```

Comment:
This line looks confusing. I doubt volatile above makes any sense. `uintptr_t` would be a more appropriate type.

Reply:
Change as suggested. It was done this way because function start_rx_dma was called with address of volatile register.
Now casting is done in placed where start_rx_dma is called instead.